### PR TITLE
NIFI-10847: Add Signal Lock for Concurrent Multiple Notify Processor

### PR DESCRIPTION
**Pull request by ZmlCode**
# Summary

[NIFI-10847](https://issues.apache.org/jira/browse/NIFI-10847)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/WaitNotifyProtocol.java [hunk 1]
```diff
@@ -181,39 +185,40 @@ public WaitNotifyProtocol(final AtomicDistributedMapCacheClient cache) {
      */
     public Signal notify(final String signalId, final Map<String, Integer> deltas, final Map<String, String> attributes)
             throws IOException, ConcurrentModificationException {
+        synchronized (pool.intern(signalId)) {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/WaitNotifyProtocol.java:188]: [Synchronizing on an interned string](https://stackoverflow.com/questions/1356277/synchronizing-on-an-interned-string) is not ideal due potential memory concerns. Some alternative approach that uses a ConcurrentMap for tracking operations might be another option.

Although Google Guava is an available dependency, using it should be avoided in favor of standard Java collections.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for working on this issue @ZmlCode. On initial evaluation, synchronizing on an interned string does not seem to be a great option, even using Google Guava's approach. It would be best to consider a solution that uses standard Java collections or locks.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for attaching the flow template to the associated Jira issue @ZmlCode.

On further review, the `Notify` Processor and this `WaitNotifyProtocol` class are functioning as designed.

There are no synchronization guarantees across multiple instances of a Processor, whether `Notify` or some other component. The `Distributed Cache Service` property provides the shared state for tracking and updating signals, delegating responsibility for concurrency outside of Processor instances.

For this reason, there is no expectation of discrete notifying processing. Introducing some type of shared static pool of signals for concurrent access management does not follow the expected behavior of the NiFi framework.

Perhaps describing the issue a bit more on the associated Jira issue could be useful, but essentially, another strategy is necessary to ensure coherent behavior when multiple Notify Processors could update the same signal.

**Comment by ZmlCode** [comment]
Thanks again.

**Pull request closed without merging**